dtls_debug.c

```diff
@@ -346,46 +389,56 @@ void dtls_dsrv_log_addr(log_t level, const char *name, const session_t *addr)
 #ifndef WITH_CONTIKI
 void
 dtls_dsrv_hexdump_log(log_t level, const char *name, const unsigned char *buf, size_t length, int extend) {
-  static char timebuf[32];
-  FILE *log_fd;
   int n = 0;
+#ifndef DTLS_CONSTRAINED_STACK
+  char message[DTLS_DEBUG_BUF_SIZE];
+#endif /* ! DTLS_CONSTRAINED_STACK */
+  size_t len;
 
   if (maxlog < (int) level)
     return;
 
-  log_fd = level <= DTLS_LOG_CRIT ? stderr : stdout;
-
-  if (print_timestamp(timebuf, sizeof(timebuf), time(NULL)))
-    fprintf(log_fd, "%s ", timebuf);
-
-  if (level <= DTLS_LOG_DEBUG)
-    fprintf(log_fd, "%s ", loglevels[level]);
-
+#ifdef DTLS_CONSTRAINED_STACK
+  dtls_mutex_lock(&static_mutex);
+#endif /* DTLS_CONSTRAINED_STACK */
   if (extend) {
-    fprintf(log_fd, "%s: (%zu bytes):\n", name, length);
-
-    while (length--) {
-      if (n % 16 == 0)
-        fprintf(log_fd, "%08X ", n);
-
-      fprintf(log_fd, "%02X ", *buf++);
+    len = snprintf(message, sizeof(message), "%s: (%zu bytes):\n", name, length);
+    while (length-- && len + 1 < sizeof(message)) {
+      if (n % 16 == 0) {
+        len += snprintf(&message[len], sizeof(message)-len, "%08X ", n);
+        if (len + 1 >= sizeof(message))
+          break;
+      }
+      len += snprintf(&message[len], sizeof(message)-len, "%02X ", *buf++);
+      if (len + 1 >= sizeof(message))
+        break;
 
       n++;
       if (n % 8 == 0) {
-        if (n % 16 == 0)
-          fprintf(log_fd, "\n");
-        else
-          fprintf(log_fd, " ");
+        if (n % 16 == 0) {
+          len += snprintf(&message[len], sizeof(message)-len, "\n");
+        } else {
+          len += snprintf(&message[len], sizeof(message)-len, " ");
+        }
       }
     }
   } else {
-    fprintf(log_fd, "%s: (%zu bytes): ", name, length);
-    while (length--)
-      fprintf(log_fd, "%02X", *buf++);
+    len = snprintf(message, sizeof(message), "%s: (%zu bytes): ", name, length);
+    while (length-- && len + 1 < sizeof(message)) {
+      len += snprintf(&message[len], sizeof(message)-len, "%02X ", *buf++);
+    }
   }
-  fprintf(log_fd, "\n");
-
-  fflush(log_fd);
+  /* len could be oversize, so re-calculate */
+  len = strlen(message);
```

Comment:
`len `at this point can be bigger than `sizeof(message)` because of previous `snprintf()`s that needed more space to write.

If `len `is untouched at this point, then random memory will get overwritten in the following lines if `message `is not big enough.  All very messy and prone to coding errors when relying on the returned length from `snprintf()`.

Reply:
```
Hm, may be that discovered a hidden issue, the code replaces the 0 in the message buffer by \n. If that the
last byte in the buffer, that smells for a un-terminated string.
```
`snprintf()` is not allowed to do that if there is insufficient space.

Reply:
```
if (len + 2 < sizeof(message))
   snprintf(&message[len], sizeof(message)-len, "\n");
else
   snprintf(&message[sizeof(message)-5], 5, " ...\n");
```

Reply:
Yes, we can reverse the order of doing the `snprintf()` by changing to your `if `test, but as I see how I now have it, it is doing exactly the same thing in terms of what is written to `message`.
````
  if (len + 1 == sizeof(message))
    /* Indicate things are truncated */
    snprintf(&message[len-5], sizeof(message)-len+5, " ...\n");
  else
    snprintf(&message[len], sizeof(message)-len, "\n");
````

Reply:
My variant obsoletes the strlen.

`len` is only used to check for overflow, and that is always `len + 1 < sizeof(message)`.
If `len` exceeds message buffer, then `len` is not used but instead the `message buffer size`.

Reply:
OK - gone with your version and pushed the code.  However, it should be &message[sizeof(message)-6] to allow for the trailing zero byte.

Reply:
Yes, I overseen the heading space.